internal/controllers/machine/machine_controller_phases.go

```diff
@@ -188,57 +139,57 @@ func (r *Reconciler) reconcileBootstrap(ctx context.Context, s *scope) (ctrl.Res
 	}
 
 	// Call generic external reconciler if we have an external reference.
-	externalResult, err := r.reconcileExternal(ctx, cluster, m, m.Spec.Bootstrap.ConfigRef)
+	obj, err := r.reconcileExternal(ctx, cluster, m, m.Spec.Bootstrap.ConfigRef)
 	if err != nil {
+		if apierrors.IsNotFound(err) {
+			if !s.machine.DeletionTimestamp.IsZero() {
+				// Tolerate bootstrap object not found when the machine is being deleted.
+				// TODO: we can also relax this and tolerate the absence of the bootstrap ref way before, e.g. after node ref is set
+				return ctrl.Result{}, nil
+			}
+			log.Info("could not find bootstrap config object, requeuing", m.Spec.Bootstrap.ConfigRef.Kind, klog.KRef(m.Spec.Bootstrap.ConfigRef.Namespace, m.Spec.Bootstrap.ConfigRef.Name))
+			// TODO: we can make this smarter and requeue only if we are before node ref is set
+			return ctrl.Result{RequeueAfter: externalReadyWait}, nil
+		}
 		return ctrl.Result{}, err
 	}
-	s.bootstrapConfig = externalResult.Result
-
-	if externalResult.RequeueAfter > 0 {
-		return ctrl.Result{RequeueAfter: externalResult.RequeueAfter}, nil
-	}
+	s.bootstrapConfig = obj
 
 	// If the bootstrap data is populated, set ready and return.
 	if m.Spec.Bootstrap.DataSecretName != nil {
 		m.Status.BootstrapReady = true
 		conditions.MarkTrue(m, clusterv1.BootstrapReadyCondition)
 		return ctrl.Result{}, nil
 	}
-	bootstrapConfig := externalResult.Result
-
-	// If the bootstrap config is being deleted, return early.
-	if !bootstrapConfig.GetDeletionTimestamp().IsZero() {
-		return ctrl.Result{}, nil
-	}
 
 	// Determine if the bootstrap provider is ready.
-	ready, err := external.IsReady(bootstrapConfig)
+	ready, err := external.IsReady(s.bootstrapConfig)
 	if err != nil {
 		return ctrl.Result{}, err
 	}
 
 	// Report a summary of current status of the bootstrap object defined for this machine.
 	conditions.SetMirror(m, clusterv1.BootstrapReadyCondition,
-		conditions.UnstructuredGetter(bootstrapConfig),
+		conditions.UnstructuredGetter(s.bootstrapConfig),
 		conditions.WithFallbackValue(ready, clusterv1.WaitingForDataSecretFallbackReason, clusterv1.ConditionSeverityInfo, ""),
```

Comment:
Does WaitingForDataSecretFallbackReason make sense during deletion?

Reply:
Turned out to be maybe the same finding as https://github.com/kubernetes-sigs/cluster-api/pull/11276#discussion_r1793200725

Reply:
it is the same, fixed